src/cascadia/TerminalControl/TSFInputControl.cpp

```diff
@@ -400,9 +393,23 @@ namespace winrt::Microsoft::Terminal::Control::implementation
         }
     }
 
+    // CoreTextRange uses int32 start/end indices whereas _inputBuffer uses unsigned size_t
+    // for a start index and length. This method safely converts the former to the latter,
+    // while making sure that out-of-range indices are clamped to the size of _inputBuffer.
+    std::pair<size_t, size_t> TSFInputControl::_convertRangeIntoBeginAndLength(CoreTextRange range) const noexcept
+    {
+        auto beg = gsl::narrow_cast<size_t>(std::max(0, range.StartCaretPosition));
+        auto end = gsl::narrow_cast<size_t>(std::max(0, range.EndCaretPosition));
+
+        end = std::min(end, _inputBuffer.size());
+        beg = std::min(beg, end);
+
+        return { beg, end - beg };
+    }

```

Comment:
FYI the `::base::ClampMin`, etc., code in this class prior to 2c922e105c6468a8c2469946d44b98390ff94f1d was not entirely correct either. For instance, it would pass `range.StartCaretPosition` to `replace()` directly.

Code like this:
```cpp
const auto textEnd = ::base::ClampMin<size_t>(range.EndCaretPosition, _inputBuffer.length());
const auto length = ::base::ClampSub<size_t>(textEnd, range.StartCaretPosition);
```
is also not correct, since the `T` in `::base::ClampSub<T>` acts just like the `T` in `std::min<T>`: It casts all parameters to `<T>` and then clamps the final value. That is, if you do `ClampSub<size_t>(123, -1)` it is the same as `ClampSub(size_t(123), size_t(-1))` which is 18446744073709551615. In other words, a negative `StartCaretPosition` will result in `length == SIZE_T_MAX` (`substr()` doesn't really care, but... eh).